### PR TITLE
Local Pairing: connection string version 2 with sharing multiple server addresses

### LINKED ISSUE #3825: Local Pairing Doesn't Work With Windows Desktop
- [Summary](#summary)
- [OS and app version](#os-and-app-version)
- [Screenshots](#screenshots)
- [Device network data](#device-network-data)

## Summary

A Windows Desktop app will not pair with another device. Windows device targets the default gateway IP of `192.168.56.1` as its preferred outbound IP, other OSes return the device's network addressable IP.

## OS and app version

```
Desktop version 0.13.1 / OS windows 10
Mobile version 1.24.0 / Pixel 7a 
```

## Screenshots

<table>
<tr>
<td></td>
<td></td>
</tr>
</table>

## Device network data

```
The computer IPv4 address is: 192.168.245.121

Where my mobile IP address is: 192.168.245.101
```

<table>
<tr>
<th>Desktop</th>
<th>Mobile</th>
</tr>
<tr>
<td><pre>
Link speed (Receive/Transmit):    1000/1000 (Mbps)
Link-local IPv6 address:    fe80::e3d5:cfcd:46a4:3ee8%3
IPv4 address:    192.168.245.121
IPv4 DNS servers:    192.168.245.1
Primary DNS suffix:    lan
DNS suffix search list:    lan
Manufacturer:    Intel
Description:    Intel(R) Ethernet Connection (2) I219-V
Driver version:    12.13.17.7
Physical address (MAC):    30-5A-3A-5A-22-A8
</pre>
<br/>
<pre>C:\Users\********>ipconfig

Windows IP Configuration

Ethernet adapter Ethernet:

   Connection-specific DNS Suffix  . : lan
   Link-local IPv6 Address . . . . . : fe80::e3d5:cfcd:46a4:3ee8%3
   IPv4 Address. . . . . . . . . . . : 192.168.245.121
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 192.168.245.1

Ethernet adapter Ethernet 2:

   Connection-specific DNS Suffix  . :
   Link-local IPv6 Address . . . . . : fe80::4016:62be:48c7:fc5c%18
   IPv4 Address. . . . . . . . . . . : 192.168.56.1
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 
</pre>
</td>
<td></td>
</tr>
</table>

### LINKED ISSUE #3884: Local Pairing - Network Address / Peer Discovery
## Network Address / Peer Discovery

Revamp the preferred network discovery process

### Self IP address acquisition

1. Ask network interface for its preferred outbound IP address
    - The process currently relied on
    - ~~Add a required a self "confirmation" ping, to ensure the given address is network addressable.~~
      - ~~This code exists [see makeOutboundCheck()](#makeoutboundcheck)~~
      - This doesn't prove anything. Don't do this to "confirm" the IP address is network accessible.
    - On fail, fall through to stage 2
2. ~~device gets a list of its network addresses.~~
   - ~~Use `netIfaces, err := net.Interfaces()` [See snippet](#get-all-network-addresses)~~
   - ~~run filter for only local private net addrs~~
   - ~~open listener on 0.0.0.0~~
   - ~~make an outbound network call to all ips~~
      - ~~call should include the addressing ip so device knows which address is network addressable~~
   - ~~record all ip addresses that are able to ping the listener.~~
      - ~~hopefully the list will only ever contain 1 IP address~~
   - ~~On fail, fall through to stage 3~~
   - ^^^ This is a **NOT** a good idea! This will not prove anything, because the device IS a part of ALL these networks and so technically could self address via any of the networks.^^^
3. Attempt discovery via UDP multicast
    - The basic code for this functionality exists [see StartSearchForLocalPairingPeers()](#startsearchforlocalpairingpeers)
    - **Question :** How do **both** devices know when to start UPD discovery?
      - Perhaps the QR code device can add a UDP discovery search flag to the `ConnectionParams`
    - no guarantee that both devices get an IP address
      - [See UDP discovery failure mitigation](#udp-discovery-failure-mitigation)

###### makeOutboundCheck()

https://github.com/status-im/status-go/blob/e620633eb745cffd54e16717ad5ac8c23e0550de/server/pairing/preflight/preflight.go#L86-L119

###### StartSearchForLocalPairingPeers()

https://github.com/status-im/status-go/blob/6cf0162877d3be58254efd1188c57b42a10e1403/mobile/status.go#L1029-L1042

###### UDP discovery failure mitigation

 - Both devices listen on a tcp port chosen by the network interface
 - Attempt UDP discovery
 - In the UDP packet include:
   - the device's listener port number
   - ensure the notifier's id is stored for later connections via tcp
 - When peer A discovers peer B, peer A pings peer B's "confirmation" listener with peer B's IP address and network info about peer A
 - peer B attempts an outbound **TLS** connection to its own "confirmation" server, with a random code, to confirm its own IP address.

###### Get all network addresses

```go
package test

import (
   "fmt"
   "net"
)

func AllNetworkAddresses() error {
   netIfaces, err := net.Interfaces()
   if err != nil {
      return err
   }

   for _, i := range netIfaces {
      addrs, err := i.Addrs()
      if err != nil {
         return err
      }

      for _, addr := range addrs {
         var ip net.IP
         switch v := addr.(type) {
         case *net.IPNet:
            ip = v.IP
         case *net.IPAddr:
            ip = v.IP
         default:
			 return fmt.Errorf("unknown ip type %s", v)
         }
         fmt.Print(ip)
      }
   }
   return nil
}
```

## Also See

- #3883

### PR DESCRIPTION
**Pull request by igor-sirotin**
Closes https://github.com/status-im/status-go/issues/3825
Closes https://github.com/status-im/status-go/issues/3884

Code has already partially reviewed in previous PRs for this branch:
- https://github.com/status-im/status-go/pull/3898
- https://github.com/status-im/status-go/pull/3904
- https://github.com/status-im/status-go/pull/3907

_____

This PR updates Local Pairing process
- server uses all available private IP addresses instead of a single IP
- to make this work, we introduce ConnectionString v2. It's backwards compatible with v1

### CODE REVIEW

File: server/ips.go [hunk 1]
```diff
@@ -20,3 +24,147 @@ func GetOutboundIP() (net.IP, error) {
...
+			ips = append(ips, niIps)
+		}
+	}
+
+	return ips, nil
+}
+
+// GetLocalAddressesForPairingServer is a high-level func
+// that returns a list of addresses to be used by local pairing server.
+func GetLocalAddressesForPairingServer() ([]net.IP, error) {
```
Comment (qfrank) [inline server/ips.go:114]: It seems we can rename some functions for clarity:

GetLocalAddresses -> getLocalAddresses
FilterAddressesForPairingServer -> filterAddressesForPairingServer
Using lowercase function names indicates which functions are intended for external usage, while uppercase names imply internal-only functions. This naming convention would help differentiate between public and private methods at a glance.

**Review by qfrank — APPROVED** [review]
Nice work!

**Comment by igor-sirotin** [comment]
Unfortunately, older versions won't see a `Not supported sync code version`. Instead, there's a general `This doesn't look like a sync code` message.

To fix this (and fix it right) we would need to implement error codes. Would be good to do that, but doesn't seem very important to implement here and now.

image

File: server/ips.go (hunk 1 above)
Comment (igor-sirotin) [reply server/ips.go:114]: Thanks for this! I was actually planning to do this, but left this for the end of implementation. And then forgot to update the names 😞

**Comment by igor-sirotin** [comment]
I was able to conduct such a test:

| Device X | Device Y |
|-|-|
| Network A 🔵  `192.168.1.33` | Network A 🔵  `192.168.1.36` |
| Network B 🟠   `192.168.2.42` |  Network C 🔴 `192.168.2.53` | 
 
Networks B 🟠  and C 🔴  are local virtual networks, though for deices X and Y they would look the same.
And there was no problem for the devices to pair 👌

**Review by Samyoul — APPROVED** [review]
Really nice! This PR covers a number of failure cases.

**Comment by Samyoul** [comment]
Just a follow up on things I've tested this PR with.

- ✅ Windows device with multiple private networks:
- ✅ Windows device with multiple valid networks
  - eg a device with multiple networks shared with both devices
- ✅ Windows device with multiple false positive networks
  - eg valid private network address that exists on multiple networks, but neither addresses are on a share network of both devices.
- ✅ Windows device with both multiple valid networks and false positive networks
- ✅ Windows device with only wifi network
- ✅ Windows device with only ethernet connection LAN
- ✅ Windows device with both wifi and ethernet connection LAN

Sample outputs:

```go
 netIPs: ([]net.IP) (len=7 cap=8) {
  (net.IP) (len=16 cap=16) 172.26.240.1,
  (net.IP) (len=16 cap=16) 172.28.192.1,
  (net.IP) (len=16 cap=16) 192.168.56.1,
  (net.IP) (len=16 cap=16) 192.168.33.1,
  (net.IP) (len=16 cap=16) 192.168.65.1,
  (net.IP) (len=16 cap=16) 192.168.159.1,
  (net.IP) (len=16 cap=16) 192.168.0.19
 },
```

```
cs3:2YAYR1henvaYDaYKVtXiknsZsb3wFg51MFeR7z2ZM:Jg9:z6VQ8GezNA1afgmR14po8AeFoHsYdtjMqANPKS14BdEJ:5gvvWVPu2MAVMTKjDeLdDAMmE5RgJUQjiRDU77ai1s5y
```

**Comment by Samyoul** [comment]
> Unfortunately, older versions won't see a `Not supported sync code version`. Instead, there's a general `This doesn't look like a sync code` message.
> 
> To fix this (and fix it right) we would need to implement error codes. Would be good to do that, but doesn't seem very important to implement here and now.
> 
> image

@igor-sirotin Hmm, those errors don't seem to come from `status-go`. Is the desktop UI returning custom errors when an error is returned from `status-go`?

But I see the value in error codes, I'll raise an issue for this.

**Comment by igor-sirotin** [comment]
> @igor-sirotin Hmm, those errors don't seem to come from status-go. Is the desktop UI returning custom errors when an error is returned from status-go?

This one is hard-coded in UI. I show it in any case of `validateConnectionString` fails.
It's not really good to show the status-go error message directly in UI (for i18n/l10n reasons). An returned error code would make it possible to show different error messages in UI.

**Comment by Samyoul** [comment]
> This one is hard-coded in UI. I show it in any case of `validateConnectionString` fails. It's not really good to show the status-go error message directly in UI (for i18n/l10n reasons). An returned error code would make it possible to show different error messages in UI.

Yes I understand, I've created the issue https://github.com/status-im/status-go/issues/3933 . So we can do localised messages etc.

**Pull request merged**